### ISSUE TITLE
downsampling for NI devices

### ISSUE THREAD

**Issue by travis-open** [AllenInstitute/MIES#1963]
Howdy,
I tried using the Downsampling panel and received the following error message:

 ```
 !!! Assertion FAILED !!!
  Message: "empty device type or number"
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
  CreateDownsamplePanel(...)#L438 [MIES_Downsample.ipf]
UpdatePanel(...)#L630 [MIES_Downsample.ipf]
UpdateDataWaves(...)#L187 [MIES_Downsample.ipf]
AppendEntries(...)#L146 [MIES_Downsample.ipf]
HW_ITC_BuildDeviceString(...)#L749 [MIES_DAC-Hardware.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2023-12-22T15:59:48-08:00
  Locked device: [Dev1]
  Current sweep: [0]
  DAQ: [DAQ_NOT_RUNNING]
  Testpulse: [TEST_PULSE_NOT_RUNNING]
  Experiment: 2023_12_22_155858 (Packed)
  Igor Pro version: 9.0.4.1 (56531)
  MIES version:
  Release_2.7_20230809-181-gbb031ee31
Date and time of last commit: 2023-11-03T14:53:28+01:00
Submodule status:
 8b830d824369ce70939e6b4e5f4b1b63fc19a79b Packages/IPNWB (remotes/origin/HEAD)
 d229d451f35e88b8c38edb8c26533acfbc3fd10f Packages/doc/doxygen-filter-ipf (heads/main)
 1b0c98b6e40dcd82fe57921e091036f1caa94992 Packages/igortest (v1.10-9-g1b0c98b)
  ################################
```
I peaked at the MIES_Downsample.ipf. Looks like downsampling for NI devices is a todo (I'm using a PCIe641). If this is a simple addition, I'd be happy to guinea pig it.  Can go other routes if this is tricky. Not at all urgent (happy holidays!).

**Comment by t-b** [comment]
Related to https://github.com/AllenInstitute/MIES/issues/31.

**Comment by timjarsky** [comment]
Happy New Year @travis-open. 

I'm pleased to know you are continuing to use MIES. The downsampling utility is ten years old, while NI device support was added in the last five years.

Can you say more about how you'd like to use this feature? Why not collect the data at a different sample rate? You may already know that it's possible to filter (NI) data using the Butterworth Sweep Formula operation.

It would be great to have users contribute code, as it's the best way to ensure MIES's long-term viability. So, if you are inclined to try, we can help.

**Comment by travis-open** [comment]
HNY @t-b and @timjarsky. I'm excited to keep using MIES. I did dip a toe in the pClamp waters before asking for new socks [a metaphor for an NI card and Igor license]. 

The reason is somewhat idiosyncratic and long-winded. I have a powered optical device that is noisy and can't be moved too far from one of the headstages. A lot of the noise is addressed by grounding, but there's some resistant high frequency noise that leads to aliasing issues if I sample at 25kHz or lower. Sampling at a high rate and then downsampling with Igor's Resample function (or sampling high and using the Butterworth Sweep Formula op) makes for nicer sweeps, but those both have trade-offs of venturing outside of the MIES ecosystem to act on MIES data or storing big files. The latter especially isn't too bad, but I was also considering the downsampling panel when I ran into the posted error message.

It's the first and only difference with regard to NI vs ITC devices that I've come across while using MIES. I super appreciate all of the work that must have gone into making the experiences so similar. I'll pursue my own noise issues and storage options, followed by seeing if I can get the downsample panel to work on NI devices. Will ping y'all if I run into any hitches (or solutions).


![Graph2](https://github.com/AllenInstitute/MIES/assets/12297639/e3291f10-fbed-452e-9797-55a121304cf9)
